R/util.R

```diff
@@ -513,3 +513,30 @@ package_root <- function(path) {
       length(grep('^Package: ', readLines(desc))) == 0) return(package_root(dir))
   dir
 }
+
+
+# retrieve package version without fear of error
+# loading namespace is ok as these packages have been or will be used
+get_package_version_string <- function(package) {
+  tryCatch(
+    as.character(getNamespaceVersion(package)),
+    error = function(e) {
+      NULL
+    }
+  )
+}
+# find all loaded packages.
+# May contain extra packages, but contains all packages used while knitting
+get_loaded_packages <- function() {
+  base_r_packages <- rownames(installed.packages(priority = 'base'))
+
+  packages <- sort(setdiff(loadedNamespaces(), base_r_packages))
+  version <- vapply(packages, get_package_version_string, character(1))
+  attached <- paste0("package:", packages) %in% search()
```

Comment:
We need to look at all loaded packages, not just the attached ones. For example, if someone does `library(shiny)`, then this will only record the version of shiny, but not dependencies like htmltools.

Another case is if they do `htmltools::div()`, it will not attach htmltools, and this code will also not record the version.